tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc

```diff
@@ -0,0 +1,573 @@
+
+#include <string>
+#include <memory>
+#include <fstream>
+#include <iostream>
+
+#include <stdio.h>
+
+#include "tensorflow/c/tf_status.h"
+
+// Implementation of a filesystem for GCS environments.
+// This filesystem will support `gs://` URI scheme.
+#define TF_ReturnIfError(status) if(TF_GetCode(status) != TF_OK) return
+
+// We can cast `google::cloud::StatusCode` to `TF_Code` because they have the same integer values.
+// See https://github.com/googleapis/google-cloud-cpp/blob/6c09cbfa0160bc046e5509b4dd2ab4b872648b4a/google/cloud/status.h#L32-L52
+static inline void TF_SetStatusFromGCSStatus(const google::cloud::Status& gcs_status, TF_Status* status) {
+	TF_SetStatus(status, static_cast<TF_Code>(gcs_status.code()), gcs_status.message().c_str());
+}
+
+static void* plugin_memory_allocate(size_t size) { return calloc(1, size); }
+static void plugin_memory_free(void* ptr) { free(ptr); }
+
+static void ParseGCSPath(const char* fname, bool object_empty_ok, char** bucket, char** object, TF_Status* status) {
+  size_t scheme_index = strcspn(fname, "://");
+  char* scheme = (char*) malloc(scheme_index + 1);
+  sprintf(scheme, "%.*s", (int)scheme_index, fname);
+  scheme[scheme_index] = '\0';
+  if(strcmp(scheme, "gs")) {
+    TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't start with 'gs://'.");
+    return;
+  }
+
+  size_t bucket_index = strcspn(fname + scheme_index + 3, "/");
+  if(!bucket_index) {
+    TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't contain a bucket name.");
+    return;
+  }
+  *bucket = (char*) malloc(bucket_index + 1);
+  sprintf(*bucket, "%.*s", (int) bucket_index, fname + scheme_index + 3);
+  (*bucket)[bucket_index] = '\0';
+
+  size_t object_index = strlen(fname + scheme_index + 3 + bucket_index + 1);
+  if(object_index == 0) {
+    if(object_empty_ok) {
+      TF_SetStatus(status, TF_OK, "");
+      *object = nullptr;
+      return;
+    }
+    else {
+      TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't contain an object name.");
+      return;
+    }
+  }
+  *object = (char*) malloc(object_index + 1);
+  sprintf(*object, "%.*s", (int) object_index, fname + scheme_index + 3 + bucket_index + 1);
+  (*object)[object_index] = '\0';
+
+  free(scheme);
+  TF_SetStatus(status, TF_OK, "");
+}
+
+static std::shared_ptr<std::fstream> CreateTempFile(char** temp_path_) {
+	*temp_path_ = (char*) malloc(L_tmpnam);
+	*temp_path_ = tmpnam(*temp_path_);
+	std::shared_ptr<std::fstream> temp_file_ = std::make_shared<std::fstream>(*temp_path_, std::fstream::binary | std::fstream::in | std::fstream::out | std::fstream::trunc);
+	//printf("TempFile: %s isopen: %d\n", *temp_path_, temp_file_->is_open());
+	return temp_file_;
+}
+
+static int64_t GetBuffer(const std::shared_ptr<std::fstream>& file, char** buffer) {
+	file->seekg(0, file->end);
+	int64_t read = file->tellg();
+	file->seekg(0, file->beg);
+	*buffer = (char*)malloc(read);
+	file->read(*buffer, read);
+	return read;
+} 
```

Comment:
This has several issues:

1. If the file is large, we allocate a large buffer on heap and this might result in OOM/DOS.
1. The file reading might not be able to read all `read` bytes in a single run, so it would have to be done in a loop.
1. There is no error checking for the file methods, these might fail.

Reply:
In fact, this is a helper function to transfer the buffer from `std::fstream` to `gcs::ObjectWriteStream` https://github.com/tensorflow/tensorflow/blob/988b3d2033a715bf3a4540ec19ec857af7e5e1d9/tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc#L196-L199

I found on stackoverflow a [way](https://stackoverflow.com/questions/4064601/copy-data-from-fstream-to-stringstream-with-no-buffer) to copy from a stream to another stream without `buffer`. I will give it a try today

Reply:
Maybe we should ask some tests like `TestUploadLargeFile` , ... which is very useful for cloud filesystem ?

Reply:
Regarding this topic, I've decided to rewrite some parts of gcs_filesystem to make it works with local temporary file instead of memory buffer. It will upload temporary file to cloud and download from cloud to temporary file, after that, it will work just like a local filesystem. I will avoid all kind of buffer as much as possible. It maybe a little slower than directly access memory buffer but it works with very large file. In order to make it possible, I would need some parts of your local filesystem plugins code ( posix and windows as well )